agcaballero/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -57,106 +96,186 @@ boolean isReplaceableBy(Availability availability) {
    * @param request the meeting request to be fulfilled (will have duration & attendees)
    */ 
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+    // The +1 is to account for last minute in day
+    Availability[] minuteAvailability = new Availability[MINUTES_IN_DAY + 1];
+
+    // this method fills the minuteAvailability array with correct availabilities and returns
+    // the max number of unavailable optional attendees for any time that day (for the attendees
+    // of this meeting)
+    int maxUnavailableOptionalAttendeesFound = fillAvailabilityArrayForDay(events, request, minuteAvailability);
+
+    for (int numberOfUnavailableOptionalAttendees = 0; 
+        numberOfUnavailableOptionalAttendees <= maxUnavailableOptionalAttendeesFound; 
+        numberOfUnavailableOptionalAttendees++) {
+      // an array list of available times 
+      ArrayList<TimeRange> availableTimes = (ArrayList<TimeRange>) findTimesBasedOnOptionalAttendeeAvailability(request, 
+        minuteAvailability, numberOfUnavailableOptionalAttendees);
+
+      // as soon as the times are found return it, bc the number of optional attendees who can't come
+      // will only increase after this point
+      if (availableTimes.size() > 0) {
+        return availableTimes;
+      }
+    }
+
+    // if there are no available times, return an empty array list
+    return new ArrayList<TimeRange>();
+  }
+  
+  /**
+   * A private helper method to determine the amount of overlap between
+   * two String Collections
+   *
+   * @return the number of elements the two Collections have in common
+   */
+  private int amountOfOverlap(Collection<String> left, Collection<String> right) {
+    int overlap = 0;
+
+    // pre-process the collections for the number of times an element occurred
+    Map<String, Integer> leftMap = countOccurrences(left);
+    Map<String, Integer> rightMap = countOccurrences(right);
+
+    // increase overlap by the number of times a given element occurred
+    // in both collections
+    for (String key: leftMap.keySet()) {
+      if(rightMap.containsKey(key)) {
+        overlap += Math.min(leftMap.get(key), rightMap.get(key));
+      }
+    }
+
+    return overlap;
+  }
```

Comment:
I think this is okay, given the requirements, but it's a bit odd... it doesn't seem like we can be sure that two "Alice"s are indeed the same Alice.

In a real use case, we'd probably want to state that these strings are some unique identifier, so that they could be safely deduped by Set.

Reply:
That makes sense. It is a bit strange. Thanks for giving me the real use case context, it's good to keep in mind.